alica_engine/include/engine/scheduler/Scheduler.h

```diff
@@ -0,0 +1,113 @@
+#pragma once
+
+#include "engine/IAlicaTimer.h"
+#include "engine/scheduler/JobQueue.h"
+
+#include <condition_variable>
+#include <mutex>
+#include <thread>
+#include <unordered_map>
+#include <utility>
+#include <vector>
+#include <atomic>
+#include <yaml-cpp/yaml.h>
+
+namespace alica
+{
+class AlicaEngine;
+
+namespace scheduler
+{
+
+template <template <class> class Queue>
+class Scheduler
+{
+public:
+    using JobId = int64_t;
+    using JobCb = std::function<void()>;
+
+private:
+    using JobQueue = Queue<std::tuple<JobId, JobCb, std::optional<AlicaTime>>>;
+
+public:
+    Scheduler(IAlicaTimerFactory& timerFactory)
+            : _running(true)
+            , _jobQueue()
+            , _mutex()
+            , _cv()
+            , _thread(&Scheduler::run, this)
+            , _nextJobId(1)
+            , _repeatableJobs()
+            , _timerFactory(timerFactory)
+    {
+    }
+
+    ~Scheduler()
+    {
+        if (_running) {
+            terminate();
+        }
+    }
+
+    JobId schedule(JobCb jobCb, std::optional<AlicaTime> repeatInterval = std::nullopt)
+    {
+        _jobQueue.emplace(_nextJobId++, std::move(jobCb), std::move(repeatInterval));
+        _cv.notify_one();
+        return _nextJobId;
+    }
+
+    void terminate()
+    {
+        _running = false;
+        _jobQueue.clear();
+
+        _cv.notify_one();
+        _thread.join();
+
+        _repeatableJobs.clear();
+    }
+
+    void stopJob(JobId jobId) { _repeatableJobs.erase(jobId); }
```

Comment:
This function should be renamed to cancelRepeatedJob(). I see a race condition here, what if the job is still in _jobQueue ? This cancel will be a no-op and scheduler will start the job later. 
One option is to have a cancelled flag for Job and not process it when the time comes.